test/provider_fallback_test.c

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2020 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stddef.h>
+#include <openssl/provider.h>
+#include <openssl/evp.h>
+#include "testutil.h"
+
+static int test_provider(OPENSSL_CTX *ctx)
+{
+    EVP_KEYMGMT *rsameth = NULL;
+    const OSSL_PROVIDER *prov = NULL;
+    int ok;
+
+    ok = TEST_true(OSSL_PROVIDER_available(ctx, "default"))
+        && TEST_ptr(rsameth = EVP_KEYMGMT_fetch(ctx, "RSA", NULL))
+        && TEST_ptr(prov = EVP_KEYMGMT_provider(rsameth))
+        && TEST_str_eq(OSSL_PROVIDER_name(prov), "default");
```

Comment:
Have you tried running this test without your patch? Wouldn't this pass anyway?, i.e. `OSSL_PROVIDER_available()` would load the fallback provider, and then unload it again. But then wouldn't the `EVP_KEYMGMT_fetch` just reload it again because we still have no providers loaded? However if we had loaded (say) the legacy provider between checking  the availability of "default" and fetching RSA, then no fallbacks will be auto-loaded after the `OSSL_PROVIDER_available` check.

Reply:
Yes I have tested.  If you look, you will see that the test commit comes first.  I also added a tag, as mentioned in https://github.com/openssl/openssl/pull/11926#issuecomment-633076426, exactly so you all could easily test for yourselves.

> i.e. `OSSL_PROVIDER_available()` would load the fallback provider, and then unload it again. But then wouldn't the `EVP_KEYMGMT_fetch` just reload it again because we still have no providers loaded?

No.  The way it's made is that there's *one* chance per library context to load the fallbacks.  As soon as an attempt has been made to activate the fallbacks, the provider store's `use_fallbacks` is zeroed, and remains that way for the lifetime of that store.
You expect something else...  so let me ask you, if you load the FIPS provider, use it for a bit, and then unload it, do you expect the fallbacks to be possible to activate again?  That's an expectation we've never talked about...

Reply:
Now that you've explained it that makes sense.